lib/Backends/OpenCL/OpenCL.cpp

```diff
@@ -1440,35 +1462,28 @@ uint64_t OpenCLFunction::copyValueFromDevice(const Value *v, void *buf) {
   return copiedBytes;
 }
 
-void OpenCLFunction::setupRuns() {
-  if (!runsSetup_) {
-    deviceBuffer_ = allocDeviceBuffer(runtimeBundle_.getConstantWeightSize() +
-                                      runtimeBundle_.getMutableWeightSize() +
-                                      runtimeBundle_.getActivationsSize());
-    size_t sizeInBytes = runtimeBundle_.getConstantWeightSize();
-    if (runtimeBundle_.getConstants()) {
-      // Issue a non-blocking command to copy the buffer to the device.
-      auto buf = runtimeBundle_.getConstants();
-      size_t valueOffset = 0;
-      cl_event event{nullptr};
-      cl_int err = clEnqueueWriteBuffer(
-          commands_, deviceBuffer_, /* blocking_write */ CL_FALSE, valueOffset,
-          sizeInBytes, buf, /* num_events_in_wait_list */ 0,
-          /* event_list */ nullptr, /* event */ clDoProfile ? &event : nullptr);
-      GLOW_ASSERT(err == CL_SUCCESS && "Unable to copy data to the device");
-      if (clDoProfile) {
-        kernelLaunches_.emplace_back(
-            KernelLaunch("copyConstantsToDevice", event));
-      }
-      // Do it!
-      clFinish(commands_);
+void OpenCLFunction::loadPlaceholders(PlaceholderBindings *bindings) {
+  size_t sizeInBytes = runtimeBundle_.getConstantWeightSize();
+  if (runtimeBundle_.getConstants()) {
+    // Issue a non-blocking command to copy the buffer to the device.
+    auto buf = runtimeBundle_.getConstants();
+    size_t valueOffset = 0;
+    cl_event event{nullptr};
+    cl_int err = clEnqueueWriteBuffer(
+        commands_, deviceBuffer_, /* blocking_write */ CL_FALSE, valueOffset,
+        sizeInBytes, buf, /* num_events_in_wait_list */ 0,
+        /* event_list */ nullptr, /* event */ clDoProfile ? &event : nullptr);
+    GLOW_ASSERT(err == CL_SUCCESS && "Unable to copy data to the device");
+    if (clDoProfile) {
+      kernelLaunches_.emplace_back(
+          KernelLaunch("copyConstantsToDevice", event));
     }
-    runsSetup_ = true;
+    // Do it!
+    clFinish(commands_);
   }
-}
 
-void OpenCLFunction::beforeRun(const PlaceholderBindings &bindings) {
-  for (auto PH : bindings.pairs()) {
+  // TODO: doesn't this copy constants twice?
```

Comment:
Hm, once when the network is added, and a second time here?

Reply:
No actually I just thought Constants were in the PlaceholderBindings for a sec, but I'm wrong.